clippy_lints/src/mut_key.rs

```diff
@@ -78,98 +78,119 @@ declare_clippy_lint! {
+            let is_keyed_type = [sym::HashMap, sym::BTreeMap, sym::HashSet, sym::BTreeSet]
+                .iter()
+                .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did()));
+            if is_keyed_type && self.is_interior_mutable_type(cx, substs.type_at(0), span) {
+                span_lint(cx, MUTABLE_KEY_TYPE, span, "mutable key type");
             }
-        },
-        _ => false,
+        }
+    }
+
+    /// Determines if a type contains interior mutability which would affect its implementation of
+    /// [`Hash`] or [`Ord`].
+    fn is_interior_mutable_type<'tcx>(&self, cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span) -> bool {
+        match *ty.kind() {
+            Ref(_, inner_ty, mutbl) => {
+                mutbl == hir::Mutability::Mut || self.is_interior_mutable_type(cx, inner_ty, span)
+            },
+            Slice(inner_ty) => self.is_interior_mutable_type(cx, inner_ty, span),
+            Array(inner_ty, size) => {
+                size.try_eval_usize(cx.tcx, cx.param_env).map_or(true, |u| u != 0)
+                    && self.is_interior_mutable_type(cx, inner_ty, span)
+            },
+            Tuple(fields) => fields.iter().any(|ty| self.is_interior_mutable_type(cx, ty, span)),
+            Adt(def, substs) => {
+                // Special case for collections in `std` who's impl of `Hash` or `Ord` delegates to
+                // that of their type parameters.  Note: we don't include `HashSet` and `HashMap`
+                // because they have no impl for `Hash` or `Ord`.
+                let is_std_collection = [
+                    sym::Option,
+                    sym::Result,
+                    sym::LinkedList,
+                    sym::Vec,
+                    sym::VecDeque,
+                    sym::BTreeMap,
+                    sym::BTreeSet,
+                    sym::Rc,
+                    sym::Arc,
+                ]
+                .iter()
+                .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did()));
+                let is_box = Some(def.did()) == cx.tcx.lang_items().owned_box();
+                if is_std_collection || is_box || self.is_arc_like(cx, def.did()) {
+                    // The type is mutable if any of its type parameters are
+                    substs.types().any(|ty| self.is_interior_mutable_type(cx, ty, span))
+                } else {
+                    !ty.has_escaping_bound_vars()
+                        && cx.tcx.layout_of(cx.param_env.and(ty)).is_ok()
+                        && !ty.is_freeze(cx.tcx.at(span), cx.param_env)
+                }
+            },
+            _ => false,
+        }
+    }
+
+    fn is_arc_like(&self, cx: &LateContext<'_>, did: hir::def_id::DefId) -> bool {
+        !self.ignore_interior_mutability.is_empty()
+            && match_any_def_paths_iter(cx, did, self.ignore_interior_mutability.iter().map(|s| s.split("::")))
+                .is_some()
     }
```

Comment:
Could use `def_path_res(..).opt_def_id()` in a `check_crate` to build a `DefIdSet` one time, similar to how `disallowed_types` does it https://github.com/rust-lang/rust-clippy/blob/191c9839f0bad1c2bfec17d55bacb94d2e83f1a1/clippy_lints/src/disallowed_types.rs#L89-L103

Then checking if a type is ignored can be a single lookup

Reply:
The problem here is that the queries used by `def_path_res` don't work on local items. It needed a good bunch of code to work around this. I'm not sure if this is absolutely correct (it might need additional code to cater for the namespace_hint), but it works in my case. If this becomes a problem, we can revisit it in a later PR.